File: openbci/offline_analysis/matlab_obci_signal_processing/ReadManager.m
classdef ReadManager < handle
    %READMANAGER Class for reading and writing channel data, tags and data
    %params
    properties
        info_source
        data_source
        tags_source        
    end
    methods
        function self=ReadManager(p_info_source,p_data_source,p_tags_source)
            %READMANAGER(p_info_source,p_data_source,p_tags_source=[])
            %   p_info_source - file_name or InfoSource
            %   p_data_source - file_name or DataSource
            %   p_tags_source - optional,file_name or TagsSource
            if ~isa(p_info_source,'InfoSource')
                p_info_source=InfoSource(p_info_source);            
            end
            self.info_source=p_info_source;
            if ~isa(p_data_source,'DataSource')
                p_data_source=FileDataSource(p_data_source,self.info_source.get_param('number_of_channels'));
            end
            self.data_source=p_data_source;
            if nargin==3 && ~isempty(p_tags_source)
                if ~isa(p_tags_source,'TagsSource')
                    p_tags_source=TagsSource(p_tags_source);
                end
                self.tags_source = p_tags_source;
            else
                self.tags_source=TagsSource();
            end
        end
        function param=get_param(self,p_param_name)
            %GET_PARAM get one param. see help InfoSource.get_param
            param=self.info_source.get_param(p_param_name);
        end

        function set_param(self,p_param_name, p_param_value)
            %SET_PARAM set one param. see help InfoSource.set_param
            self.info_source.set_param(p_param_name, p_param_value);
        end

        function start_ts=get_start_timestamp(self)
            %GET_START_TIMESTAMP gets start timestamp of this stream
            start_ts=0;
        end
        function params=get_params(self)
            %GET_PARAMS get all params. see help InfoSource.get_params
            params=self.info_source.get_params();
        end         
        function set_params(self,p_params)
            %SET_PARAMS(p_params) 
            %   p_params= struct
            %   Set params. see help InfoSource.set_params
            self.info_source.set_params(p_params)
        end
        function samples=get_samples(self,p_from,p_len)
            %GET_SAMPLES(p_from,p_len) get samples. 
            %   see help DataSource.get_samples
            %   when invoked without arguments, all samples are cached in
            %   memory, so it is faster in the future.
            if nargin==1
                samples=self.data_source.get_samples();
                if ~isa(self.data_source,'MemoryDataSource')
                    self.data_source=MemoryDataSource(samples);
                end
            end
            if nargin<3; p_len=inf; end
            if nargin<2; p_from=0; end                 
            samples=self.data_source.get_samples(p_from,p_len);
        end
        
        function iter_samples(self,p_from,p_len)
            %ITER_SAMPLES iter through samples
            %   see help DataSource.iter_samples
            if nargin<3; p_len=inf; end
            if nargin<2; p_from=0; end            
            self.data_source.iter_samples(p_from,p_len)          
        end
        function samples=next_samples(self)
            %NEXT_SAMPLES iter through samples
            %   see help DataSource.next_samples
            samples=self.data_source.next_samples();
        end
        function set_samples(self,p_samples,p_channel_names)
            %SET_SAMPLES(p_samples,p_channel_names)
            %   p_samples - matrix(num_channels,num_samples)
            %   p_channel_names - array of strings of length of
            %   num_channels
            %Usage:
            % set_samples([1,2,3;4,5,6],{'A','B'})
            s=size(p_samples);
            if s(1)>0 && (~exist('p_channel_names','var')||length(p_channel_names)~=s(1))
                throw(MException('ReadManager:WrongChannelNames','Wrong number of channel names'));
            end
            self.data_source=MemoryDataSource();
            self.data_source.set_samples(p_samples);
            self.info_source.set_param('channels_names',p_channel_names);
            self.info_source.set_param('number_of_channels',length(p_channel_names));
            self.info_source.set_param('number_of_samples',s(2))  ;          
        end
        function ch_samples=get_channel_samples(self,p_ch_name,p_from,p_len)
            %GET_CHANNEL_SAMPLES(p_ch_name, p_from=[],p_len=inf) - gets all
            %samples fron given channel
            %   p_ch_name - channel index or channel name as string
            %   p_from - from which sample to start
            %   p_len - how many samples
            if nargin<4; p_len=inf; end
            if nargin<3; p_from=0; end            
            samples=self.get_samples(p_from,p_len);
            if isinteger(p_ch_name)
                ch_ind=p_ch_name;
            else
                ch_ind=find(ismember(self.get_param('channels_names'),p_ch_name)==1);
            end
            ch_samples=samples(ch_ind,:);
        end
        function tags=get_tags(self,p_tag_type,p_from,p_len,p_func)
            %GET_TAGS(p_tag_type,p_from,p_len,p_func) get filtered tags.
            %   see help TagsSource.get_tags
           if nargin<5; p_func=@(x) true;end
            if nargin<4; p_len=inf;end
            if nargin<3; p_from=0;end
            if nargin<2; p_tag_type=[];end
            tags = self.tags_source.get_tags(p_tag_type,p_from,p_len,p_func);
        end
        
        function iter_tags(self,p_tag_type,p_from,p_len,p_func)
            %ITER_TAGS(p_tag_type,p_from,p_len,p_func) get filtered tags.
            %   see help TagsSource.iter_tags
            if nargin<4; p_func=@(x) true;end
            if nargin<3; p_len=inf;end
            if nargin<2; p_from=0;end
            if nargin<1; p_tag_type=[];end
            self.tags_source.iter_tags(p_tag_type,p_from,p_len,p_func);
        end
        function tag=next_tag(self)
            %NEXT_TAG() get filtered tags.
            %   see help TagsSource.next_tag
            tag=self.tags_source.next_tag();
        end
        function set_tags(self,p_tags,p_first_ts)
            %SET_TAGS(p_tags,p_first_timestamp) set tags.
            %   see help TagsSource.set_tags
            default('p_first_ts',0)
            self.tags_source.set_tags(p_tags,p_first_ts)
        end
        function save_to_file(self,p_dir,p_name)
            %SAVE_TO_FILE(p_dir,p_name) save date from this ReadManager to
            %   files 
            %   p_dir - directory path
            %   p_name - prefix for file names .obci.dat, .obci.info,
            %   .obci.tags
            
            if ~isa(self.data_source,'MemoryDataSource')
                [~]=self.get_samples();
            end
            path=[p_dir '/' p_name];
            self.data_source.save_to_file([path '.obci.dat']);
            self.info_source.save_to_file([path '.obci.info']);
            self.tags_source.save_to_file([path '.obci.tags']);            
        end
    end
end

File: openbci/offline_analysis/obci_signal_processing/read_manager.py

# -*- coding: utf-8 -*-
#!/usr/bin/env python
#
# OpenBCI - framework for Brain-Computer Interfaces based on EEG signal
# Project was initiated by Magdalena Michalska and Krzysztof Kulewski
# as part of their MSc theses at the University of Warsaw.
# Copyright (C) 2008-2009 Krzysztof Kulewski and Magdalena Michalska
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# Author:
#     Mateusz Kruszyński <mateusz.kruszynski@gmail.com>
#
import sys, struct, time, numpy, copy, os.path

from signal import read_data_source
from signal import read_info_source
from tags import read_tags_source

from signal import data_file_proxy
from signal import info_file_proxy
from tags import tags_file_writer

import obci_signal_processing_logging as logger
LOGGER = logger.get_logger("read_manager", "info")

class ReadManager(object):
    """A class responsible for reding openbci file format.
    Public interface:
    - start_reading() - open info and data file,
    - get_next_value() - get next value from data file,
    - get_param(param_name) - get param_name parameter from info file.

    Wanna be able to read a new parameter 'new_param'?
    1. Register reading function in self._create_info_tags_control() under 'new_param' key.
    2. Implement the function (it should be considered as class private function, not callable from outside; 
    the function should return a value for 'new_param' request).
    3. Call get_param('new_param') every time you want to get the param.
    """
    def __init__(self, p_info_source, p_data_source, p_tags_source):
        """Just remember info file path and data file path."""
        try:
            ''+p_info_source
            LOGGER.debug("Got info source file path.")
            self.info_source = read_info_source.FileInfoSource(p_info_source)
        except TypeError:
            LOGGER.debug("Got info source object.")
            self.info_source = p_info_source

        try:
            ''+p_data_source
            LOGGER.debug("Got data source file path.")
            self.data_source = read_data_source.FileDataSource(
                p_data_source,
                int(self.info_source.get_param('number_of_channels'))
                )
        except TypeError:
            LOGGER.debug("Got data source object.")
            self.data_source = p_data_source

        try:
            ''+p_tags_source
            LOGGER.debug("Got tags source file path.")
            self.tags_source = read_tags_source.FileTagsSource(p_tags_source)
        except TypeError:
            LOGGER.debug("Got tags source object.")
            self.tags_source = p_tags_source 

    def __deepcopy__(self, memo):
        info_source = copy.deepcopy(self.info_source)
        tags_source = copy.deepcopy(self.tags_source)
        samples_source = copy.deepcopy(self.data_source)
        return ReadManager(info_source, samples_source, tags_source)

    def save_to_file(self, p_dir, p_name):
        tags = self.get_tags()
        params = self.get_params()

        path = os.path.join(p_dir, p_name)
        #store tags
        tags_writer = tags_file_writer.TagsFileWriter(path+'.obci.tags')
        for tag in tags:
            tags_writer.tag_received(tag)

        #store info
        info_writer = info_file_proxy.InfoFileWriteProxy(path+'.obci.info')
        info_writer.set_attributes(params)

        #store data
        data_writer = data_file_proxy.DataFileWriteProxy(path+'.obci.dat')
        for sample in self.iter_samples():
            for d in sample:
                data_writer.data_received(d)

        tags_writer.finish_saving(0)        
        info_writer.finish_saving()
        data_writer.finish_saving()


    def get_samples(self, p_from=None, p_len=None):
        """Return a two dimensional array of signal values.
        if p_reload then refresh the file, otherwise use cached values."""
        
        return self.data_source.get_samples(p_from, p_len)

    def get_channel_samples(self, p_ch_name, p_from=None, p_len=None):
        """Return an array of values for channel p_ch_name, or
        raise ValueError exception if there is channel with that name."""
        ch_ind = self.get_param('channels_names').index(p_ch_name) #TODO error
        return self.get_samples(p_from, p_len)[ch_ind]

    def set_samples(self, p_samples, p_channel_names, p_copy=False):
        try:
            dim = p_samples.ndim
        except AttributeError:
            raise Exception("Samples not a numpy array!")
        if (dim != 2):
            raise Exception("Samples must be a 2-dim numpy array!")
        num_of_channels = p_samples.shape[0]
        num_of_samples = p_samples.shape[1]
        
        if (len(p_channel_names) != num_of_channels):
            raise Exception("Number of channels names is different from number of channels in samples!")
        
        self.set_param('channels_names', p_channel_names)
        self.set_param('number_of_channels', num_of_channels)
        self.set_param('number_of_samples', num_of_samples)
        self.data_source.set_samples(p_samples, p_copy)


    def get_tags(self, p_tag_type=None, p_from=None, p_len=None, p_func=None):
        """Return all tags of type tag_type, or all types if tag_type is None."""

        if self.tags_source is None:
            return []
        else:
            return self.tags_source.get_tags(p_tag_type, p_from, p_len, p_func)
    def set_tags(self, p_tags):
        if self.tags_source is not None:
            self.tags_source.set_tags(p_tags)
        else:
            self.tags_source = read_tags_source.MemoryTagsSource(p_tags)

    def get_param(self, p_param_name):
        """Return parameter value for p_param_name.
        Raise NoParameter exception if p_param_name 
        parameters was not found."""
        return self.info_source.get_param(p_param_name)

    def get_params(self):
        return self.info_source.get_params()

    def set_param(self, p_key, p_value):
        self.info_source.set_param(p_key, p_value)

    def set_params(self, p_params):
        self.info_source.update_params(p_params)

    def reset_params(self):
        self.info_source.reset_params()

    def iter_tags(self, p_tag_type=None, p_from=None, p_len=None, p_func=None):
        if self.tags_source is None:
            return 

        for tag in self.tags_source.get_tags(p_tag_type, p_from, p_len, p_func):
            yield tag
        
    def iter_samples(self):
        for s in self.data_source.iter_samples():
            yield s
